File: lib/models/scan_result.dart
import 'dart:convert';
import 'package:intl/intl.dart';
import 'threat.dart';

class ScanResult {
  final String id;
  final String type;
  final DateTime startTime;
  DateTime? endTime;
  int totalScanned;
  List<Threat> threatsFound;
  String? userId;
  double progress; // Yüzde olarak ilerleme durumu

  ScanResult({
    required this.id,
    required this.type,
    required this.startTime,
    this.endTime,
    this.totalScanned = 0,
    this.threatsFound = const [],
    this.userId,
    this.progress = 0,
  });

  // Tarama süresi (saniye cinsinden)
  int get duration {
    if (endTime == null) {
      // Eğer tarama hala devam ediyorsa şu ana kadar geçen süreyi hesapla
      return DateTime.now().difference(startTime).inSeconds;
    }
    // Tarama tamamlanmışsa başlangıç ve bitiş arası süreyi hesapla
    return endTime!.difference(startTime).inSeconds;
  }

  // Tarama durumu
  String get status {
    if (endTime == null) {
      return 'RUNNING';
    }
    return threatsFound.isEmpty ? 'COMPLETED' : 'WARNING';
  }

  // JSON'dan nesne oluşturma
  factory ScanResult.fromJson(Map<String, dynamic> json) {
    // Tehdit listesini oluştur
    List<Threat> threats = [];
    if (json['threatsFound'] != null) {
      threats = (json['threatsFound'] as List)
          .map((threatJson) => Threat.fromJson(threatJson))
          .toList();
    }

    // Tarih ve saat formatını parse et
    final dateFormat = DateFormat("yyyy-MM-ddTHH:mm:ss");
    DateTime? endTime;
    
    if (json['endTime'] != null && json['endTime'] != '') {
      endTime = dateFormat.parse(json['endTime']);
    }

    return ScanResult(
      id: json['_id'] ?? json['id'],
      type: json['type'],
      startTime: dateFormat.parse(json['startTime']),
      endTime: endTime,
      totalScanned: json['totalScanned'] ?? 0,
      threatsFound: threats,
      userId: json['userId'],
      progress: (json['progress'] ?? 0).toDouble(),
    );
  }

  // JSON'a dönüştürme
  Map<String, dynamic> toJson() {
    final dateFormat = DateFormat("yyyy-MM-ddTHH:mm:ss");
    
    return {
      'id': id,
      'type': type,
      'startTime': dateFormat.format(startTime),
      'endTime': endTime != null ? dateFormat.format(endTime!) : null,
      'totalScanned': totalScanned,
      'threatsFound': threatsFound.map((threat) => threat.toJson()).toList(),
      'userId': userId,
      'progress': progress,
    };
  }

  // Taramayı tamamla
  void complete({
    required int totalScanned,
    required List<Threat> threatsFound,
  }) {
    this.endTime = DateTime.now();
    this.totalScanned = totalScanned;
    this.threatsFound = threatsFound;
    this.progress = 100;
  }

  // İlerleme durumunu güncelle
  void updateProgress(double progress, {int? itemsScanned, List<Threat>? newThreats}) {
    this.progress = progress;
    
    if (itemsScanned != null) {
      this.totalScanned = itemsScanned;
    }
    
    if (newThreats != null) {
      this.threatsFound = newThreats;
    }
  }

  // Yeni tarama sonucu oluştur
  static ScanResult createNew(String type, String? userId) {
    // Benzersiz ID oluştur
    final timestamp = DateTime.now().millisecondsSinceEpoch;
    final id = 'scan_$timestamp';
    
    return ScanResult(
      id: id,
      type: type,
      startTime: DateTime.now(),
      userId: userId,
    );
  }
}

File: models/scanResult.js
/**
 * ScanResult Model
 * Tarama sonuçlarını temsil eden model sınıfı
 */

const { v4: uuidv4 } = require('uuid');
const database = require('../config/database');
const Threat = require('./threat');

/**
 * ScanResult model
 * Tarama sonuçlarını temsil eden model sınıfı
 */
class ScanResult {
  /**
   * Yeni bir tarama sonucu nesnesi oluşturur
   * @param {string} id - Tarama ID'si
   * @param {string} type - Tarama tipi (QUICK, FULL, WIFI, QR, vb.)
   * @param {Date} startTime - Tarama başlangıç zamanı
   * @param {Date|null} endTime - Tarama bitiş zamanı (null ise devam ediyor)
   * @param {number} totalScanned - Taranan öğe sayısı
   * @param {Threat[]} threatsFound - Bulunan tehditler dizisi
   * @param {string|null} userId - Taramayı yapan kullanıcı ID'si (null ise sistem taraması)
   */
  constructor(id, type, startTime, endTime = null, totalScanned = 0, threatsFound = [], userId = null) {
    this.id = id || uuidv4();
    this.type = type;
    this.startTime = startTime || new Date();
    this.endTime = endTime;
    this.totalScanned = totalScanned;
    this.threatsFound = threatsFound || [];
    this.userId = userId;
  }

  /**
   * Tarama sonucunu JSON formatına dönüştürür
   * @returns {object} JSON formatında tarama sonucu
   */
  toJSON() {
    return {
      id: this.id,
      type: this.type,
      startTime: this.startTime,
      endTime: this.endTime,
      totalScanned: this.totalScanned,
      threatsFound: this.threatsFound.map(threat => 
        typeof threat === 'string' ? threat : threat.toJSON()
      ),
      userId: this.userId,
      status: this.getStatus(),
      duration: this.getDuration()
    };
  }

  /**
   * Tarama durumunu kontrol eder
   * @returns {string} - Tarama durumu (RUNNING, COMPLETED, FAILED)
   */
  getStatus() {
    if (!this.endTime) return 'RUNNING';
    return 'COMPLETED';
  }

  /**
   * Tarama süresini hesaplar (milisaniye cinsinden)
   * @returns {number} - Tarama süresi
   */
  getDuration() {
    if (!this.endTime) {
      return Date.now() - this.startTime.getTime();
    }
    return this.endTime.getTime() - this.startTime.getTime();
  }

  /**
   * Taramayı tamamlar
   * @param {number} totalScanned - Toplam taranan öğe sayısı
   * @param {Threat[]} threatsFound - Bulunan tehditler
   */
  complete(totalScanned, threatsFound) {
    this.endTime = new Date();
    this.totalScanned = totalScanned;
    this.threatsFound = threatsFound;
  }

  /**
   * Tarama sonucunu veritabanına kaydeder
   * @returns {Promise<boolean>} - Kaydetme işlemi başarılı ise true döner
   */
  async save() {
    try {
      const db = database.getDb();
      if (!db) return false;

      const scanResults = db.collection('scanResults');
      
      const threatIds = [];
      
      // Tehdit objeleri varsa, bunları önce kaydet ve ID'lerini al
      for (const threat of this.threatsFound) {
        if (typeof threat === 'string') {
          threatIds.push(threat);
        } else {
          await threat.save();
          threatIds.push(threat.id);
        }
      }
      
      const scanData = {
        _id: this.id,
        type: this.type,
        startTime: this.startTime,
        endTime: this.endTime,
        totalScanned: this.totalScanned,
        threatsFound: threatIds,
        userId: this.userId
      };

      // Tarama sonucunun daha önce kaydedilip kaydedilmediğini kontrol et
      const existingScan = await scanResults.findOne({ _id: this.id });
      
      if (existingScan) {
        // Tarama sonucu zaten var, güncelle
        await scanResults.updateOne({ _id: this.id }, { $set: scanData });
      } else {
        // Yeni tarama sonucu ekle
        await scanResults.insertOne(scanData);
      }
      
      return true;
    } catch (error) {
      console.error('Tarama sonucu kaydedilirken hata:', error);
      return false;
    }
  }

  /**
   * ID'ye göre tarama sonucu bulur
   * @param {string} id - Tarama ID'si
   * @returns {Promise<ScanResult|null>} - Bulunan tarama sonucu, yoksa null
   */
  static async findById(id) {
    try {
      const db = database.getDb();
      if (!db) return null;

      const scanResults = db.collection('scanResults');
      const scanData = await scanResults.findOne({ _id: id });
      
      if (!scanData) return null;
      
      // Tehditleri getir
      const threatIds = scanData.threatsFound || [];
      let threats = [];
      
      if (threatIds.length > 0) {
        const threatsCollection = db.collection('threats');
        const threatsData = await threatsCollection.find({ _id: { $in: threatIds } }).toArray();
        
        threats = threatsData.map(threatData => new Threat(
          threatData._id,
          threatData.name,
          threatData.type,
          threatData.description,
          threatData.severity,
          threatData.filePath,
          threatData.isCleaned,
          threatData.detectionDate
        ));
      }
      
      return new ScanResult(
        scanData._id,
        scanData.type,
        scanData.startTime,
        scanData.endTime,
        scanData.totalScanned,
        threats,
        scanData.userId
      );
    } catch (error) {
      console.error('ID ile tarama sonucu aranırken hata:', error);
      return null;
    }
  }

  /**
   * Kullanıcıya ait tarama sonuçlarını getirir
   * @param {string} userId - Kullanıcı ID'si
   * @param {number} limit - Maksimum sonuç sayısı
   * @returns {Promise<ScanResult[]>} - Tarama sonuçları dizisi
   */
  static async findByUserId(userId, limit = 10) {
    try {
      const db = database.getDb();
      if (!db) return [];

      const scanResults = db.collection('scanResults');
      const scanDataList = await scanResults.find({ userId })
        .sort({ startTime: -1 })
        .limit(limit)
        .toArray();
      
      const results = [];
      
      for (const scanData of scanDataList) {
        // Tehditleri getir
        const threatIds = scanData.threatsFound || [];
        let threats = [];
        
        if (threatIds.length > 0) {
          const threatsCollection = db.collection('threats');
          const threatsData = await threatsCollection.find({ _id: { $in: threatIds } }).toArray();
          
          threats = threatsData.map(threatData => new Threat(
            threatData._id,
            threatData.name,
            threatData.type,
            threatData.description,
            threatData.severity,
            threatData.filePath,
            threatData.isCleaned,
            threatData.detectionDate
          ));
        }
        
        results.push(new ScanResult(
          scanData._id,
          scanData.type,
          scanData.startTime,
          scanData.endTime,
          scanData.totalScanned,
          threats,
          scanData.userId
        ));
      }
      
      return results;
    } catch (error) {
      console.error('Kullanıcı ID\'si ile tarama sonuçları aranırken hata:', error);
      return [];
    }
  }

  /**
   * Son tarama sonuçlarını getirir
   * @param {number} limit - Maksimum sonuç sayısı
   * @returns {Promise<ScanResult[]>} - Tarama sonuçları dizisi
   */
  static async findRecent(limit = 10) {
    try {
      const db = database.getDb();
      if (!db) return [];

      const scanResults = db.collection('scanResults');
      const scanDataList = await scanResults.find()
        .sort({ startTime: -1 })
        .limit(limit)
        .toArray();
      
      const results = [];
      
      for (const scanData of scanDataList) {
        // Tehditleri getir
        const threatIds = scanData.threatsFound || [];
        let threats = [];
        
        if (threatIds.length > 0) {
          const threatsCollection = db.collection('threats');
          const threatsData = await threatsCollection.find({ _id: { $in: threatIds } }).toArray();
          
          threats = threatsData.map(threatData => new Threat(
            threatData._id,
            threatData.name,
            threatData.type,
            threatData.description,
            threatData.severity,
            threatData.filePath,
            threatData.isCleaned,
            threatData.detectionDate
          ));
        }
        
        results.push(new ScanResult(
          scanData._id,
          scanData.type,
          scanData.startTime,
          scanData.endTime,
          scanData.totalScanned,
          threats,
          scanData.userId
        ));
      }
      
      return results;
    } catch (error) {
      console.error('Son tarama sonuçları getirilirken hata:', error);
      return [];
    }
  }

  /**
   * Tarama sayısını getirir
   * @param {object} filter - Filtreleme seçenekleri
   * @returns {Promise<number>} - Tarama sayısı
   */
  static async count(filter = {}) {
    try {
      const db = database.getDb();
      if (!db) return 0;

      const scanResults = db.collection('scanResults');
      return await scanResults.countDocuments(filter);
    } catch (error) {
      console.error('Tarama sayısı getirilirken hata:', error);
      return 0;
    }
  }

  /**
   * Simüle edilmiş tarama sonucu oluşturur (Test ve demo amacıyla)
   * @param {string} type - Tarama tipi
   * @param {string|null} userId - Kullanıcı ID'si
   * @returns {ScanResult} - Simüle edilmiş tarama sonucu
   */
  static createSimulatedScan(type, userId = null) {
    const id = `scan${uuidv4().substring(0, 8)}`;
    const startTime = new Date();
    const totalScanned = Math.floor(Math.random() * 500) + 50; // 50-550 arası
    
    // Rastgele tehdit sayısı (0-5 arası)
    const threatCount = Math.floor(Math.random() * 6);
    const threatsFound = Threat.getRandomThreats(threatCount);
    
    const scanResult = new ScanResult(
      id,
      type,
      startTime,
      null, // endTime
      0, // totalScanned (tarama devam ediyor)
      threatsFound,
      userId
    );
    
    return scanResult;
  }
}

module.exports = ScanResult;